x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java

```diff
@@ -135,14 +136,19 @@ public CacheHelper getReaderCacheHelper() {
      */
     FieldSubsetReader(LeafReader in, CharacterRunAutomaton filter) throws IOException {
         super(in);
-        ArrayList<FieldInfo> filteredInfos = new ArrayList<>();
-        for (FieldInfo fi : in.getFieldInfos()) {
-            if (filter.run(fi.name)) {
+        this.fieldInfos = filter(in.getFieldInfos(), filter);
+        this.filter = filter;
+    }
+
+    // pkg-private for testing
+    static FieldInfos filter(FieldInfos fieldInfos, CharacterRunAutomaton filter) {
```

Comment:
nit: I know naming is hard, but we now have 4 filter methods. It would be nice to make their use case a bit more distinguishable. What you think about `filterPermittedFields`?